internal/core/software_renderer.rs

```diff
@@ -1591,8 +1591,63 @@ impl<'a, T: ProcessScene> crate::item_rendering::ItemRenderer for SceneBuilder<'
     }
 
     fn draw_text_input(&mut self, text_input: Pin<&crate::items::TextInput>, _: &ItemRc) {
-        text_input.geometry();
-        // TODO
+        let string = text_input.text();
+        let geom = LogicalRect::new(LogicalPoint::default(), text_input.geometry().size_length());
+        if !self.should_draw(&geom) {
+            return;
+        }
+
+        let font_request = text_input.font_request(&self.window.window_adapter());
+
+        let color = self.alpha_color(text_input.color().color());
+        let max_size = (geom.size.cast() * self.scale_factor).cast();
+
+        // Clip glyphs not only against the global clip but also against the Text's geometry to avoid drawing outside
+        // of its boundaries (that breaks partial rendering and the cast to usize for the item relative coordinate below).
+        // FIXME: we should allow drawing outside of the Text element's boundaries.
+        let physical_clip = if let Some(logical_clip) = self.current_state.clip.intersection(&geom)
+        {
+            logical_clip.cast() * self.scale_factor
+        } else {
+            return; // This should have been caught earlier already
+        };
+        let offset = self.current_state.offset.to_vector().cast() * self.scale_factor;
+
+        let font = fonts::match_font(&font_request, self.scale_factor);
+
+        match font {
+            fonts::Font::PixelFont(pf) => {
+                let paragraph = TextParagraphLayout {
+                    string: &string,
+                    layout: fonts::text_layout_for_font(&pf, &font_request, self.scale_factor),
+                    max_width: max_size.width_length(),
+                    max_height: max_size.height_length(),
+                    horizontal_alignment: text_input.horizontal_alignment(),
+                    vertical_alignment: text_input.vertical_alignment(),
+                    wrap: text_input.wrap(),
+                    overflow: TextOverflow::Clip,
+                    single_line: text_input.single_line(),
+                };
+
+                self.draw_text_paragraph(paragraph, physical_clip, offset, color);
+            }
+            #[cfg(feature = "software-renderer-systemfonts")]
+            fonts::Font::VectorFont(vf) => {
+                let paragraph = TextParagraphLayout {
+                    string: &string,
+                    layout: fonts::text_layout_for_font(&vf, &font_request, self.scale_factor),
```

Comment:
nitpick:  It looks like only this line is different from the two match arms.  perhaps worth to simplify this by only doing the layout in the match, and move the common code outside of the match.

Reply:
Yes but both layouts have incompatible types what would not work with the match.